### DISCUSSION TITLE
getCellValue in QTable @request

### DISCUSSION THREAD

**Discussion by Divensky** [quasarframework/quasar#17106] [General - Components / Directives / etc]
Hi! 

The docs on `@request` event of `q-table` state that the params of `requestProp` include a **required** param `getCellValue` . The index.d.ts file does not give it a question mark as an optional parameter, so it must be required: 
```
    /**
     * Function to get a cell value
     * @param col Column name from column definitions
     * @param row The row object
     * @returns Parsed/Processed cell value
     */
    getCellValue: (col: any, row: any) => any;
```

Two questions:
1. I discovered that in fact I can omit this param and my request works just fine. Is it a quirk or is it meant to work like this? 

Here is my working code:
```
// Handle built-in @request events
    const onRequest: OnRequest = (requestProp: RequestProp) => {
      emit('server-request', requestProp);
    };
    
// Constuct a requestProp and fire @request events
   const changeRowsPerPage = (value: number) => {
      pagination.value.rowsPerPage = value;
      const requestProp = {
        pagination: pagination.value,
      };
      emit('server-request', requestProp);
    };

// The event handler in the parent handles 'server-request' successfully in either case
```

2. Could anyone help get a broader perspective on `getCellValue` prop? I've searched and did not find anywhere an example of how it was meant to be used. 

I see that `@request` gets triggered when any pagination button is pressed and includes an updated `pagination` object. Also, `@request` gets emitted when a filter is pressed (as long as `v-model` on `:filter` is configured) with an updated `filter` object. 

I did not find such use cases for the `getCellValue` prop. 

Note: There is a relevant discussion on [Q-table server requests](https://github.com/quasarframework/quasar/discussions/13960), but it does not encompass any of my Qs above.

**Comment by yusufkandemir** [comment]
> I discovered that in fact I can omit this param and my request works just fine. Is it a quirk or is it meant to work like this?

If you try to access `getCellValue` in a `@server-request` handler, you will see it won't work. So, it might seem like it's working until it's not, but it's entirely dependent on your usage. QTable has `getCellValue` internally, so it won't run into issues by itself.

> Could anyone help get a broader perspective on getCellValue prop? I've searched and did not find anywhere an example of how it was meant to be used.

It can be called as `getCellValue(col, row)` as shown in the API card. Basically, it uses the `field` and `format` definitions in the column definition(provided via `columns` prop), then returns you the related formatted cell value. For example:
```js
// column definition that you defined in the columns prop
const col = {
  field: (row) => row.foo.bar,
  format: (value) => value.toFixed(2),
}

const row = {
  foo: {
    bar: 15.2345
  }
}

getCellValue(col, row) // => '15.23'
```
Here is the source code of `getCellValue` definition:
https://github.com/quasarframework/quasar/blob/dbb51486a41f697ef9b76aedbbd0404984574212/ui/src/components/table/QTable.js#L528-L531

> I see that `@request` gets triggered when any pagination button is pressed and includes an updated pagination object. Also, `@request` gets emitted when a filter is pressed (as long as v-model on :filter is configured) with an updated filter object.

Yes, and that's why you shouldn't emit `server-request` in `changeRowsPerPage`, as `pagination.value.rowsPerPage = value;` would trigger the `@request` event, which would end up calling your `const onRequest = ...`.

**Comment by Divensky** [comment]
Thanks a lot! Much clearer. Is there a user interaction that natively triggers an update of `getCellValue` object in a @request? 

> Yes, and that's why you shouldn't emit `server-request` in `changeRowsPerPage`, as `pagination.value.rowsPerPage = value;` would trigger the `@request` event, which would end up calling your `const onRequest = ...`

No, I observed that `pagination.value.rowsPerPage = value` does not by itself trigger the `@request` event. The pagination does update (the console shows new values) but no update of the table rows occurs. 

I guess that's because my `changeRowsPerPage()` is not triggered by a native pagination button. It is an extra button that I added. The native one emits `@request` all right but my own button would not until I wrote the above. 

Or maybe another part of this is that the built-in `@request` is meant to contact the server in the direct parent of `q-table`. My use case is different: I have a wrapper around `q-table`, and it is the parent of the wrapper that contacts the server.
